### PR TITLE
rest, endpoints: Reduce code duplication

### PR DESCRIPTION
**Pull request by vasiliy-ul**
**What this PR does / why we need it**:

The PR introduces common functions: `formatURI`, `httpGetRequestHandler` and applies them accross the code to reduce duplication. No functional changes.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/virt-api/rest/subresource.go [hunk 1]
```diff
@@ -135,6 +135,30 @@ func (app *SubresourceAPIApp) putRequestHandler(request *restful.Request, respon
 	}
 }
 
+func (app *SubresourceAPIApp) getRequestHandler(request *restful.Request, response *restful.Response, validate validation, getURL URLResolver, v interface{}) {
```
Comment (EdDev) [inline pkg/virt-api/rest/subresource.go:138]: Any chance the name can be improved?
It does not "get" anything (sounds like a getter), and the name "Handler" does not really hint what it does exactly.

Sorry for not suggesting any other name here, I just do not really understand what this piece of code does eventually.

**Review by EdDev — COMMENTED** [review]
Thank you!

File: pkg/virt-api/rest/subresource.go (hunk 1 above)
Comment (vasiliy-ul) [reply pkg/virt-api/rest/subresource.go:138]: Its a handler of HTTP GET request. I named it after similar handler for HTTP PUT: 

https://github.com/kubevirt/kubevirt/blob/f3c6c7df391b134d689da19b526e4c6f298c0170/pkg/virt-api/rest/subresource.go#L121

Maybe smth like `httpGetRequestHandler`?

Comment (EdDev) [reply pkg/virt-api/rest/subresource.go:138]: Yea, this suggested name is less confusing. Thanks!

Comment (vasiliy-ul) [reply pkg/virt-api/rest/subresource.go]: Updated to `httpGetRequestHandler`

**Review by EdDev — APPROVED** [review]
Thank you!

**Comment by vasiliy-ul** [comment]
@EdDev , @brybacki , @rmohr Could you take one more look? I added a function `decodeRequestBody` in `subresource.go` (the code was duplicted in several places). Since the PR is not merged yet I though it would be better to just update it.

File: pkg/virt-api/rest/subresource.go [hunk 2]
```diff
@@ -1108,22 +1045,16 @@ func (app *SubresourceAPIApp) addVolumeRequestHandler(request *restful.Request,
 		return
 	}
 
-	opts := &v1.AddVolumeOptions{}
-	if request.Request.Body != nil {
-		defer request.Request.Body.Close()
```
Comment (vasiliy-ul) [inline pkg/virt-api/rest/subresource.go:1113]: according to https://github.com/golang/go/commit/2ede818ae0761fc57dc1b52b4c8673933a3ff3fa there is no need to close the request body.

File: pkg/virt-api/rest/subresource.go [hunk 3]
```diff
@@ -135,6 +136,44 @@ func (app *SubresourceAPIApp) putRequestHandler(request *restful.Request, respon
...
+	if err := json.Unmarshal([]byte(resp), &v); err != nil {
+		log.Log.Reason(err).Error("error unmarshalling response")
+		response.WriteError(http.StatusInternalServerError, err)
+		return
+	}
+
+	response.WriteEntity(v)
+}
+
+func (app *SubresourceAPIApp) decodeRequestBody(request *restful.Request, response *restful.Response, bodyStruct interface{}) bool {
```
Comment (EdDev) [inline pkg/virt-api/rest/subresource.go:163]: I like less the fact that this one returns a boolean.

Also, the is a subtle change, where you the body is passed originally as an address/pointer, and now by value.

**Review by EdDev — COMMENTED** [review]
This PR and its single commit do too much IMO.
Some is explained/reasoned as a PR comment and some implicit.
Having a commit for each functional improvement would be better, allowing also to share the reasoning in detail.

The addition to the PR now is also problematic in my view, as an "approve" has been given by a maintainer to the prev changes, and not to this new thing. I guess canceling it and re-requesting it is one option to go.
But note, that generally speaking, it is faster to have small PR/s over loading it with more changes, just  because there is a potential dragging that can pop up on one of the changes.

**Comment by vasiliy-ul** [comment]
> This PR and its single commit do too much IMO. Some is explained/reasoned as a PR comment and some implicit. Having a commit for each functional improvement would be better, allowing also to share the reasoning in detail.
> 
> The addition to the PR now is also problematic in my view, as an "approve" has been given by a maintainer to the prev changes, and not to this new thing. I guess canceling it and re-requesting it is one option to go. But note, that generally speaking, it is faster to have small PR/s over loading it with more changes, just because there is a potential dragging that can pop up on one of the changes.

@EdDev , makes sense. Probably better if I revert the added changes and restore the PR to the original state. Sorry for mixing things here.

**Comment by vasiliy-ul** [comment]
Restored the initial state. Will address the new issues separately. Sorry for confusion.

File: pkg/virt-api/rest/subresource.go (hunk 3 above)
Comment (vasiliy-ul) [reply pkg/virt-api/rest/subresource.go]: > Also, there is a subtle change, where the body is passed originally as an address/pointer, and now by value.

I think that should be fine. The `Decode` method also accepts ` interface{}` and the actual `bodyStruct` passed here is a pointer, e.g. ` bodyStruct := &v1.MigrateOptions{}`.

**Review by EdDev — APPROVED** [review]

File: pkg/virt-api/rest/subresource.go (hunk 3 above)
Comment (EdDev) [reply pkg/virt-api/rest/subresource.go]: Yea, it was a double pointer.

**Pull request merged**